include/af/memory.h

```diff
@@ -0,0 +1,307 @@
+AFAPI af_err af_memory_manager_get_active_device_id(af_memory_manager handle,
+                                                    int* id);
+
+AFAPI af_err af_memory_manager_native_alloc(af_memory_manager handle,
+                                            void** ptr, size_t size);
+
+AFAPI af_err af_memory_manager_native_free(af_memory_manager handle, void* ptr);
+
+AFAPI af_err af_memory_manager_get_max_memory_size(af_memory_manager handle,
+                                                   size_t* size, int id);
+
+#ifdef __cplusplus
+}
+#endif  // __cplusplus
+
+#ifdef __cplusplus
+namespace af {
+
+class AFAPI memory_manager {
+    af_memory_manager m_;
+
+    typedef af_memory_manager_initialize_fn InitializeFn;
+    typedef af_memory_manager_shutdown_fn ShutdownFn;
+    typedef af_memory_manager_alloc_fn AllocFn;
+    typedef af_memory_manager_allocated_fn AllocatedFn;
+    typedef af_memory_manager_unlock_fn UnlockFn;
+    typedef af_memory_manager_garbage_collect_fn GarbageCollectFn;
+    typedef af_memory_manager_print_info_fn PrintInfoFn;
+    typedef af_memory_manager_usage_info_fn UsageInfoFn;
+    typedef af_memory_manager_user_lock_fn UserLockFn;
+    typedef af_memory_manager_user_unlock_fn UserUnlockFn;
+    typedef af_memory_manager_is_user_locked_fn IsUserLockedFn;
+    typedef af_memory_manager_get_mem_step_size_fn GetMemStepSizeFn;
+    typedef af_memory_manager_get_max_bytes_fn GetMaxBytesFn;
+    typedef af_memory_manager_get_max_buffers_fn GetMaxBuffersFn;
+    typedef af_memory_manager_set_mem_step_size_fn SetMemStepSizeFn;
+    typedef af_memory_manager_check_memory_limit CheckMemoryLimitFn;
+    typedef af_memory_manager_add_memory_management AddMemoryManagementFn;
+    typedef af_memory_manager_remove_memory_management RemoveMemoryManagementFn;
+
+   public:
+    memory_manager();
+    memory_manager(af_memory_manager p);
+    ~memory_manager();
+#if AF_COMPILER_CXX_RVALUE_REFERENCES
+    memory_manager(memory_manager&& other);
+    memory_manager& operator=(memory_manager&& other);
+#endif
+    af_memory_manager get() const;
+
+    void registerInitialize(InitializeFn fn);
+    void registerShutdown(ShutdownFn fn);
+    void registerAlloc(AllocFn fn);
+    void registerAllocated(AllocatedFn fn);
+    void registerUnlock(UnlockFn fn);
+    void registerGarbageCollect(GarbageCollectFn fn);
+    void registerPrintInfo(PrintInfoFn fn);
+    void registerUsageInfo(UsageInfoFn fn);
+    void registerUserLock(UserLockFn fn);
+    void registerUserUnlock(UserUnlockFn fn);
+    void registerIsUserLocked(IsUserLockedFn fn);
+    void registerGetMemStepSize(GetMemStepSizeFn fn);
+    void registerGetMaxBytes(GetMaxBytesFn fn);
+    void registerGetMaxBuffers(GetMaxBuffersFn fn);
+    void registerSetMemStepSize(SetMemStepSizeFn fn);
+    void registerCheckMemoryLimit(CheckMemoryLimitFn fn);
+    void registerAddMemoryManagement(AddMemoryManagementFn fn);
+    void registerRemoveMemoryManagement(RemoveMemoryManagementFn fn);
```

Comment:
It is probably more explicit with `register*` notation. I wonder if we can template the register function instead of repeating it so many times.

Reply:
Thought about this - would be nice to have. I couldn't find a way to dispatch to the correct call of `af_memory_mananger_set_[function name]` from `register<T>` - seemed like it involves some pretty hideous templating with the preprocessor. If you have any ideas, do let me know

Reply:
If it is not too much of development effort, we should simply it now so that any new `registerMethod` that we want to add later becomes a simple one line edit in future.